ggml/src/ggml-cuda/mmf.cuh

```diff
@@ -57,31 +57,37 @@ static __global__ void mul_mat_f(
     T * tile_xy = (T *) compute_base + threadIdx.y*(tile_A::I * tile_k_padded);
 
     if constexpr (has_ids) {
-        __shared__ int has_any;
-        if (threadIdx.y == 0) {
-            int local_has_any = 0;
-            for (int j = threadIdx.x; j < cols_per_block; j += warp_size) {
-                int slot = -1;
-                for (int k = 0; k < nchannels_dst; ++k) {
-                    const int idv = ids[j*stride_row_id + k*stride_col_id];
-                    if (idv == expert_idx) {
-                        slot = k;
-                        break;
+        int found = 0;
+
+        for (int j = threadIdx.y; j < cols_per_block; j += nwarps) {
+            const int32_t * __restrict__ id_row = ids + j*stride_row_id;
+
+            if (threadIdx.x == 0) {
+                slot_map[j] = -1;
+            }
+
+            for (int k_base = 0; k_base < nchannels_dst; k_base += warp_size) {
```

Comment:
Take a look at how `iex_used` is determined in `mmq_ids_helper`. You can exploit the fact that the expert you're looking for always appears either 0 or 1 times in `ids`. So if you condition the write to `slot_map` on the expert being found you have a guarantee that only a single thread will write to that location.